recipes/libiconv/all/conanfile.py

```diff
@@ -55,99 +63,72 @@ def configure(self):
-            lt_compile = tools.unix_path(os.path.join(build_aux_path, "compile"))
-            lt_ar = tools.unix_path(os.path.join(build_aux_path, "ar-lib"))
-            env_vars.update({
-                "CC": "{} {} -nologo".format(lt_compile, cc),
-                "CXX": "{} {} -nologo".format(lt_compile, cxx),
-                "LD": "link",
-                "STRIP": ":",
-                "AR": "{} {}".format(lt_ar, lib),
-                "RANLIB": ":",
-                "NM": "dumpbin -symbols"
-            })
-            env_vars["win32_target"] = "_WIN32_WINNT_VISTA"
-
-        if not tools.cross_building(self) or is_msvc(self) or self._is_clang_cl:
-            rc = None
-            if self.settings.arch == "x86":
-                rc = "windres --target=pe-i386"
-            elif self.settings.arch == "x86_64":
-                rc = "windres --target=pe-x86-64"
-            if rc:
-                env_vars["RC"] = rc
-                env_vars["WINDRES"] = rc
-        if self._use_winbash:
-            env_vars["RANLIB"] = ":"
-
-        with tools.vcvars(self.settings) if (is_msvc(self) or self._is_clang_cl) else tools.no_op():
-            with tools.chdir(self._source_subfolder):
-                with tools.environment_append(env_vars):
-                    yield
-
-    @functools.lru_cache(1)
-    def _configure_autotools(self):
-        host = None
-        build = None
         if is_msvc(self) or self._is_clang_cl:
-            build = False
-            if self.settings.arch == "x86":
-                host = "i686-w64-mingw32"
-            elif self.settings.arch == "x86_64":
-                host = "x86_64-w64-mingw32"
+            cc, lib, link = self._msvc_tools
+            build_aux_path = os.path.join(self.source_folder, "build-aux")
+            lt_compile = unix_path(self, os.path.join(build_aux_path, "compile"))
+            lt_ar = unix_path(self, os.path.join(build_aux_path, "ar-lib"))
+            env.define("CC", f"{lt_compile} {cc} -nologo")
+            env.define("CXX", f"{lt_compile} {cc} -nologo")
+            env.define("LD", f"{link}")
+            env.define("STRIP", ":")
+            env.define("AR", f"{lt_ar} {lib}")
+            env.define("RANLIB", ":")
+            env.define("NM", "dumpbin -symbols")
+            env.define("win32_target", "_WIN32_WINNT_VISTA")
 
-        autotools = AutoToolsBuildEnvironment(self, win_bash=tools.os_info.is_windows)
 
-        configure_args = []
-        if self.options.shared:
-            configure_args.extend(["--disable-static", "--enable-shared"])
-        else:
-            configure_args.extend(["--enable-static", "--disable-shared"])
+        tc.generate(env)
 
-        if (self.settings.compiler == "Visual Studio" and tools.Version(self.settings.compiler.version) >= "12") or \
-           self.settings.compiler == "msvc":
-            autotools.flags.append("-FS")
+        env = VirtualBuildEnv(self)
+        env.generate()
```

Comment:
should be called before AutotoolsToolchain, otherwise it can defeats msvc env var tricks above.

Reply:
Good to know! 👍

Reply:
@SpaceIm @uilianries this is still called _after_ AutotoolsToolchain,
can this recipe be fixed up to be done the "proper" way?
I am trying to use this recipe as a model for upgrading libsodium, and its not good if this recipe isn't "gold standard" (or suggests another to follow)

Reply:
Actually, how does this even work properly?  "env" variable name is reused...
first as   env = tc.environment() and then as env=VirtualBuildEnv(self).
how should it be done "before AutotoolsToolchain" ?

Reply:
Use the paclage templates as a refernce point, https://github.com/conan-io/conan-center-index/blob/master/docs/package_templates/autotools_package/all/conanfile.py#L91 

we are trying to focus on keeping those up to spec